lib/msal-common/src/response/ResponseHandler.ts

```diff
@@ -0,0 +1,131 @@
+import { Account } from "../auth/Account";
+import { ProtocolUtils } from "../utils/ProtocolUtils";
+import { ICrypto } from "../crypto/ICrypto";
+import { ICacheStorage } from "../cache/ICacheStorage";
+import { TokenResponse } from "./TokenResponse";
+import { PersistentCacheKeys } from "../utils/Constants";
+import { ClientAuthError } from "../error/ClientAuthError";
+
+/**
+ * Class that handles response parsing.
+ */
+export class ResponseHandler {
+    private clientId: string;
+    private cacheStorage: ICacheStorage;
+    private cacheManager: CacheHelpers;
+    private cryptoObj: ICrypto;
+
+    constructor(clientId: string, cacheStorage: ICacheStorage, cacheManager: CacheHelpers, cryptoObj: ICrypto) {
+        this.clientId = clientId;
+        this.cacheStorage = cacheStorage;
+        this.cacheManager = cacheManager;
+        this.cryptoObj = cryptoObj;
+    }
+
+    public setResponseIdToken(originalResponse: TokenResponse, idTokenObj: IdToken) : TokenResponse {
+        if (!originalResponse) {
+            return null;
+        } else if (!idTokenObj) {
+            return originalResponse;
+        }
+    
+        const exp = Number(idTokenObj.claims.exp);
+        if (exp && !originalResponse.expiresOn) {
+            originalResponse.expiresOn = new Date(exp * 1000);
+        }
+    
+        return {
+            ...originalResponse,
+            idToken: idTokenObj.rawIdToken,
+            idTokenClaims: idTokenObj.claims,
+            uniqueId: idTokenObj.claims.oid || idTokenObj.claims.sub,
+            tenantId: idTokenObj.claims.tid,
+        };
+    }
+    
+    public createTokenResponse(serverTokenResponse: ServerAuthorizationTokenResponse, state: string): TokenResponse {
+        let tokenResponse: TokenResponse = {
+            uniqueId: "",
+            tenantId: "",
+            tokenType: "",
+            idToken: null,
+            idTokenClaims: null,
+            accessToken: "",
+            refreshToken: "",
+            scopes: [],
+            expiresOn: null,
+            account: null,
+            userRequestState: ""
+        };
+        // Set consented scopes in response
+        const requestScopes = ScopeSet.fromString(serverTokenResponse.scope, this.clientId, false);
+        tokenResponse.scopes = requestScopes.asArray();
+    
+        // Retrieve current id token object
+        let idTokenObj: IdToken;
+        const cachedIdToken: string = this.cacheStorage.getItem(PersistentCacheKeys.ID_TOKEN);
+        if (serverTokenResponse.id_token) {
+            idTokenObj = new IdToken(serverTokenResponse.id_token, this.cryptoObj);
+            tokenResponse = this.setResponseIdToken(tokenResponse, idTokenObj);
+        } else if (cachedIdToken) {
+            idTokenObj = new IdToken(cachedIdToken, this.cryptoObj);
+            tokenResponse = this.setResponseIdToken(tokenResponse, idTokenObj);
+        } else {
+            // TODO: No account scenario?
```

Comment:
Throw exception maybe?

Reply:
Need to do more investigation, i will tackle this in a separate PR.